### ISSUE TITLE
Extend kernel capabilities in kata-static archive

### ISSUE THREAD

**Issue by jzaehrin** [kata-containers/kata-containers#3146]
**Which feature do you think can be improved?**

The content of the `kata-static` archive.

**How can it be improved?**

Adding `kernels` with specific capabilities like `SEV` inside the `kata-static` archive. This would avoid having to compile these kernels locally when you want to use specific features.

**Additional Information**

Currently, I use the `-x sev` option on the `build-kernel.sh` script to generate the SEV compatible Linux kernel.

**Comment by fidencio** [comment]
@jzaehrin, this aligns with an interesting discussion that @eadamsintel and I had last week.

There's a problem on this that we couldn't figure out how to solve, or what would be the best approach to solve.  Imagine we add the kernel with SEV as part of the kata-static, and then another one supporting TDX, and then another one supporting ths or that.  We'll end up with a way too big set of binaries as part of our image, and that may not be desired.

We also discussed the possibility of having a fully bloated kernel (hey @jodh-intel, we also discussed this), and maybe distribute that advertising that won't have the best memory consumption or the fastest boot time, but that would serve the purpose for users to test kata-containers with their environment.  Maybe that's an option, dunno.

There's also the fact that we should avoid shipping things we don't test, and so far we don't test kata + SEV (or at least there's no public CI about that).

Anyways, I am glad you brought this up and I think we should carefully evaluate the approach we take, so we can expand this for our users without shooting ourselves in the foot wrt the maintainability.  We'll address this, and any kind of input is super welcome!

**Comment by eadamsintel** [comment]
Myself and @fidencio discussed this but in the context of `kata-deploy`. This would apply to the entire build and validation process. We thought use case specific kernels optimized for memory, boot speed, or functionality made the most sense. The tricky part is how do we differentiate and distribute so the user experience is good? I have not looked at the results of the Kata survey but understanding how customers use `kata-deploy` would really help.  My inclination is that `kata-deploy` isn't used for large scale deployments, but rather for small deployments or for testing.  If that is the case then having a fully featured container image with multiple kernels would make sense.  If `kata-deploy` is used in a lot of production instances then having container images full of kernels that someone might not use would take up unnecessary space. In that scenario having multiple container tags for for different use cases might make sense. However, would someone ever want two different optimized kernels from two different Docker tags in the same cluster? If so then installing that through `kata-deploy` would be tricky.  We discussed this for at least an hour the other day and realized that a lot of people will have a lot of opinions on how to do this.

My mind is easily swayed, but right now I am leaning towards a single `kata-deploy` tag for each architecture that has every type of optimized kernel and rootfs in it using `runtimeClass` to separate out which optimized kernel is used.  You could have a GPU optimized kernel, FAAS optimized kernel, fully featured unoptimized kernel, etc..  This would lead to bigger binaries installed on each persons system.  In my opinion nothing should be bundled into `kata-deploy` or the static tarball unless it has been through validation.

**Comment by jzaehrin** [comment]
From a distribution maintainer point of view, the current kata-static structure is very rigid and offers too many components. In our case, for example, we are not interested in getting the qemu build in kata-static.

It seems to me that it would be interesting to separate entities to avoid downloading big archives for a limited part of it. And in this sense, it would be possible to propose experimental archives containing kernels with features not yet tested. However, it seems to me that GitHub is no longer the right platform to provide so many archives.

Unfortunately, I did not use kata-deploy, so it is difficult for me to comment on this subject. However, adding all the features to the kernel is as much impacting for the user ?

**Comment by fidencio** [comment]
> From a distribution maintainer point of view, the current kata-static structure is very rigid and offers too many components. In our case, for example, we are not interested in getting the qemu build in kata-static.

I get that and, to be totally fair, from a distribution maintainer point of view one should only care about the kata-containers code itself (as in, runtime and agent), and then use as much as possible from the distro, including using the kernel, qemu, and rootfs / initrd for a distro based experience.  That's we ended up doing when I was working on the Fedora / CentOS / RHCOS side.

> It seems to me that it would be interesting to separate entities to avoid downloading big archives for a limited part of it. And in this sense, it would be possible to propose experimental archives containing kernels with features not yet tested. However, it seems to me that GitHub is no longer the right platform to provide so many archives.

For now distros can just drop the files they don't want.  It's a cheap workaround but may serve the purpose till we figure out where we could store things.

> Unfortunately, I did not use kata-deploy, so it is difficult for me to comment on this subject. However, adding all the features to the kernel is as much impacting for the user ?

It impacts in the memory consumed by each pod, for instance.  The more bloated the kernel for the guest VM, the more memory consumed just to have the pod up and running.

**Comment by jodh-intel** [comment]
> We also discussed the possibility of having a fully bloated kernel

Yes, I think the "full fat" kernel is the way to go as it lowers the barrier to entry and will work in all possible scenarios. It also aligns with what distros provide of course. There is the question of whether that default kernel should be build monolithically or use modules (which Kata supports), but honestly, it probably doesn't matter either way since this is the kernel which "just works", **not** an optimised kernel as would be used for serious use-cases.

I've added this topic to the Architecture Committee meeting for later today (we may not get around to discussing it today though given the other items on the agenda):

- https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs

**Comment by jzaehrin** [comment]
> I get that and, to be totally fair, from a distribution maintainer point of view one should only care about the kata-containers code itself (as in, runtime and agent), and then use as much as possible from the distro, including using the kernel, qemu, and rootfs / initrd for a distro based experience. That's we ended up doing when I was working on the Fedora / CentOS / RHCOS side.

Thanks, I hadn't seen that way, I'll look into it.

> For now distros can just drop the files they don't want. It's a cheap workaround but may serve the purpose till we figure out where we could store things.

Yes totally, I will follow the evolution of this in the future.

Thank you for your answers.

**Comment by jodh-intel** [comment]
I've written a small presentation on a full fat kernel concept. We ran out of time in the AC meeting today, but hopefully I'll be able to present it early in the New Year.

The presentation is here: https://gist.github.com/jodh-intel/b9e0d241b6bb1700bc0074dadc3c99b8
(see the comment at the end if you want to view it as an actual presentation!)

**Comment by jodh-intel** [comment]
I've created a new issue to avoid hijacking this one: #3399.
